util/encrypt/aes.go

```diff
@@ -120,26 +122,49 @@ func PKCS7Unpad(data []byte, blockSize int) ([]byte, error) {
 
 // AESEncryptWithECB encrypts data using AES with ECB mode.
 func AESEncryptWithECB(str, key []byte) ([]byte, error) {
-	cb, err := aes.NewCipher(key)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	blockSize := cb.BlockSize()
-	// The str arguments can be any length, and padding is automatically added to
-	// str so it is a multiple of a block as required by block-based algorithms such as AES.
-	// This padding is automatically removed by the AES_DECRYPT() function.
-	data, err := PKCS7Pad(str, blockSize)
-	if err != nil {
-		return nil, err
-	}
-	crypted := make([]byte, len(data))
-	ecb := newECBEncrypter(cb)
-	ecb.CryptBlocks(crypted, data)
-	return crypted, nil
+	return aesEncrypt(str, key, func(block cipher.Block) cipher.BlockMode {
+		return newECBEncrypter(block)
+	})
 }
 
 // AESDecryptWithECB decrypts data using AES with ECB mode.
 func AESDecryptWithECB(cryptStr, key []byte) ([]byte, error) {
+	return aesDecrypt(cryptStr, key, func(block cipher.Block) cipher.BlockMode {
+		return newECBDecrypter(block)
+	})
+}
+
+// DeriveKeyMySQL derives the encryption key from a password in MySQL algorithm.
+// See https://security.stackexchange.com/questions/4863/mysql-aes-encrypt-key-length.
+func DeriveKeyMySQL(key []byte, blockSize int) []byte {
+	rKey := make([]byte, blockSize)
+	rIdx := 0
+	for _, k := range key {
+		if rIdx == blockSize {
+			rIdx = 0
+		}
+		rKey[rIdx] ^= k
+		rIdx++
+	}
+	return rKey
+}
+
+// AESEncryptWithCBC encrypts data using AES with CBC mode.
+func AESEncryptWithCBC(str, key []byte, iv []byte) ([]byte, error) {
+	return aesEncrypt(str, key, func(block cipher.Block) cipher.BlockMode {
+		return cipher.NewCBCEncrypter(block, iv)
+	})
+}
+
+// AESDecryptWithCBC decrypts data using AES with CBC mode.
+func AESDecryptWithCBC(cryptStr, key []byte, iv []byte) ([]byte, error) {
+	return aesDecrypt(cryptStr, key, func(block cipher.Block) cipher.BlockMode {
+		return cipher.NewCBCDecrypter(block, iv)
+	})
+}
+
+// aesDecrypt decrypts data using AES.
+func aesDecrypt(cryptStr, key []byte, build blockModeBuild) ([]byte, error) {
```

Comment:
It's better to pass `cipher.BlockMode` instead of `blockModeBuild`

Reply:
AESEncryptWithCBC/AESDecryptWithCBC is similar with AESEncryptWithECB/AESDecryptWithECB with different cipher.BlockMode, CBC needs `iv` but ECB does not, so blockModeBuild is the hook method to build cipher.BlockMode.

Reply:
IMHO, if we can remove this hook, pass the required `cipher.BlockMode` directly, the code would be easier to read and maintain.